sdgym/_dataset_utils.py

```diff
@@ -0,0 +1,105 @@
+"""Utility functions for handling datasets."""
+
+import json
+import logging
+from pathlib import Path
+from zipfile import ZipFile
+
+import numpy as np
+import pandas as pd
+
+LOGGER = logging.getLogger(__name__)
+
+
+def _parse_numeric_value(value, dataset_name, field_name, target_type=float):
+    """Generic parser for numeric values with logging and NaN fallback."""
+    try:
+        return target_type(value)
+    except (ValueError, TypeError):
+        LOGGER.info(
+            f"Could not cast {field_name} '{value}' to {target_type.__name__} for dataset "
+            f"'{dataset_name}' defaulting to NaN."
+        )
+        return np.nan
+
+
+def _get_dataset_subset(data, metadata_dict, modality):
+    """Limit the size of a dataset for faster evaluation or testing.
+
+    This function reduces a dataset to a smaller subset by restricting the number
+    of rows and columns to 1000 rows and 10 columns. It ensures that essential
+    columns—such as sequence indices and keys in sequential datasets—are always retained.
+
+    Args:
+        data (pd.DataFrame):
+            The dataset to be reduced.
+        metadata_dict (dict):
+            A dictionary containing the dataset's metadata.
+        modality (str):
+            The dataset modality. Must be one of: ``'single_table'``, ``'sequential'``.
+
+    Returns:
+        tuple[pd.DataFrame, dict]:
+            A tuple containing:
+            - The reduced dataset as a DataFrame.
+            - The updated metadata dictionary reflecting any removed columns.
+
+    Raises:
+        ValueError:
+            If the provided modality is ``'multi_table'``.
+    """
+    if modality == 'multi_table':
+        raise ValueError('limit_dataset_size is not supported for multi-table datasets.')
+
+    max_rows, max_columns = (1000, 10)
+    tables = metadata_dict.get('tables', {})
+    mandatory_columns = []
+    for table_name, table_info in tables.items():
+        columns = table_info.get('columns', {})
+        if modality == 'sequential':
+            seq_index = table_info.get('sequence_index')
+            seq_key = table_info.get('sequence_key')
+            mandatory_columns = [col for col in (seq_index, seq_key) if col]
+
+        optional_columns = [col for col in columns if col not in mandatory_columns]
+
+        # If we have too many columns, drop extras but never mandatory ones
+        if len(columns) > max_columns:
+            keep_count = max_columns - len(mandatory_columns)
```

Comment:
Could mandatory_columns not be greater than max_columns? If so keep_count is negative, so when you do `optional_columns[:keep_count]` in the line below it will add unwanted columns.